src/sentry/utils/snuba.py

```diff
@@ -593,55 +568,148 @@ def raw_query(start, end, groupby=None, conditions=None, filter_keys=None,
     if start > end:
         raise QueryOutsideGroupActivityError
 
-    kwargs.update({
+    query_params.kwargs.update({
         'from_date': start.isoformat(),
         'to_date': end.isoformat(),
-        'groupby': groupby,
-        'conditions': conditions,
-        'aggregations': aggregations,
+        'groupby': query_params.groupby,
+        'conditions': query_params.conditions,
+        'aggregations': query_params.aggregations,
         'project': project_ids,
-        'granularity': rollup,  # TODO name these things the same
+        'granularity': query_params.rollup,  # TODO name these things the same
     })
-    kwargs = {k: v for k, v in six.iteritems(kwargs) if v is not None}
+    kwargs = {k: v for k, v in six.iteritems(query_params.kwargs) if v is not None}
 
     kwargs.update(OVERRIDE_OPTIONS)
+    return kwargs, forward, reverse
+
+
+class SnubaQueryParams(object):
```

Comment:
Would be really nice if this method had an `execute()` method that called raw_query(). This would let the rest of the application use this builder class to define a query and then execute it without having to import multiple bits from utils/snuba.

Reply:
That's an interesting idea... I think it should call `bulk_raw_query` and pass itself rather than `raw_query`, since that would just involve recreating the object again.

Could potentially have a collection class of `SnubaQueryParams` as well that has an execute that can be used to bulk execute a bunch of queries as well, but potentially unnecessary since we can just call `bulk` directly with the list.

Will think about this a bit.

Reply:
@markstory Thought about this, and I'm not sure it will simplify things much.

If you want to make one query, you can jump import `raw_query` and call that. This the main case that `SnubaQueryParams.execute` would handle, so I'm not sure that it's saving much work there?

If you want to make multiple queries, then you need to import both `SnubaQueryParams` and  `bulk_raw_query` so that you can build the param list. If we had `SnubaQueryBuilder.add_query` then we could potentially make this a single import. I'm not sure if it's worth the extra abstraction though, but I'm open to it.